# Stercoralis RNAseq Browser Shiny App

## ---- Libraries ----
suppressPackageStartupMessages({
    library(shiny)
    library(shinyWidgets)
    library(htmltools)
    library(shinythemes)
    library(DT)
    library(tidyverse)
    library(limma) # differential gene expression using linear modeling
    library(edgeR)
    library(gt) 
    library(plotly)
    library(shinycssloaders)
    library(magrittr)
    library(ggthemes)
    library(gplots)
    library(heatmaply)
    library(RColorBrewer)
    source("Server/calc_DEG_tbl.R")
    source("Server/theme_Publication.R")
})

## ---- Background ----
# Load pre-calculated data generated by S_stercoralis_RNAseq_ReAnalysis.R
#load(file = "./SsLifeStageComparisons_subset")
#load(file = "./SsLifeStageComparisons_TopHits")
# Import a variance-stabilized DEGList created by voom transformation command.
# Outputs: E = normalized CPMexpression values on the log2 scale
load(file = "./vDEGList")

# Parse vDEGList into a tibble containing Log2CPM information
Log2CPM<-v.DEGList.filtered.norm$E %>%
    as_tibble(rownames = "geneID")%>%
    setNames(nm = c("geneID", 
                    as.character(v.DEGList.filtered.norm$targets$group))) %>%
    pivot_longer(cols = -geneID,
                 names_to = "life_stage", 
                 values_to = "log2CPM") %>%
    group_by(geneID, life_stage)

diffGenes.df <- v.DEGList.filtered.norm$E %>%
    as_tibble(rownames = "geneID", .name_repair = "unique")

# Fit a linear model to the data ----
fit <- lmFit(v.DEGList.filtered.norm, v.DEGList.filtered.norm$design)


# Set Expression threshold values for plotting and saving DEGs ----
adj.P.thresh <- 0.05
lfc.thresh <- 0

## ---- UI ----

# Application title ----
#

ui <- fluidPage(
    #titlePanel("Strongyloides stercoralis RNAseq Browser"),
    #source('UI/SsRNAseqBrowser_dashboard-ui.R', local = T)$value
    source('UI/SsRNAseqBrowser_navbar-ui.R', local = T)$value
    
)


# Define server logic  ----
server <- function(input, output, session) {
    
    vals <- reactiveValues(genelist = NULL,
                           gene_of_interest = NULL,
                           highlight.df = NULL,
                           comparison = NULL,
                           targetStage = NULL,
                           contrastStage = NULL,
                           multipleCorrection = NULL,
                           results = NULL,
                           ebFit = NULL)
    
    ## GW: Single Gene Expression Plot ----
    
    observeEvent(input$goGW,{
        vals$genelist <- NULL
        vals$gene_of_interest <- NULL
        vals$highlight.df <- NULL
        vals$comparison <- NULL
        vals$targetStage <- NULL
        vals$contrastStage <- NULL
        vals$multipleCorrection <- NULL
        vals$results <- NULL
        vals$ebFit <- NULL
        updateSelectInput(session,"selectContrast_GW", selected = "", choices = NULL)
        updateSelectInput(session,"selectTarget_GW", selected = "", choices = NULL)
        updateSelectInput(session, "displayedComparison_GW", selected = "", choices = NULL)
    }, ignoreInit = TRUE)
    
    parse_ids <- eventReactive(input$goGW,{
        if (isTruthy(input$idtext)){
            genelist <- input$idtext %>%
                gsub(" ", "", ., fixed = TRUE) %>%
                str_split(pattern = ",") %>%
                unlist() %>%
                as_tibble_col(column_name = "geneID")
        } else if (isTruthy(input$loadfile)){
            file <- input$loadfile
            ext <- tools::file_ext(file$datapath)
            validate(need(ext == "csv", "Please upload a csv file"))
            genelist <- read.csv(file$datapath, 
                                 header = FALSE, 
                                 colClasses = "character", 
                                 strip.white = T) %>%
                as_tibble() %>%
                pivot_longer(cols = everything(), 
                             values_to = "geneID") %>%
                dplyr::select(geneID)
        } 
        
        vals$genelist <- genelist
    })
    
    ### GW: Generate Responsive Selection for Gene to Display ----
    output$geneDisplaySelection_GW <- renderUI({
        panel(
            heading = tagList(h5(shiny::icon("fas fa-filter"),
                                 "Step 1B: Select Gene to Display")),
            status = "default",
            selectInput("displayedGene",
                        NULL, 
                        c("All Genes", vals$genelist$geneID))
        )
    })
    
    ### GW: Plot Gene Expression by Life Stage ----
    output$CPM <- renderPlot({
        parse_ids()
        # Select gene to display
        if (isTruthy(input$displayedGene)){
            if (input$displayedGene == "All Genes"){
                vals$gene_of_interest <- vals$genelist$geneID
            } else vals$gene_of_interest <- input$displayedGene
        } else {vals$gene_of_interest <- vals$genelist[[1,1]]}
        
        
        if (length(vals$gene_of_interest) == 1) {
            # Plot Log2CPM values for an individual gene
            gene_vals <- Log2CPM %>%
                dplyr::filter(geneID == vals$gene_of_interest)
            
            p<- ggplot(gene_vals) + 
                aes(x = life_stage, y = log2CPM, fill = life_stage) +
                geom_boxplot(show.legend = F, alpha = 0.7) +
                labs(y="log2 CPM expression", x = "Life Stage",
                     title=paste("Log2 Counts per Million (CPM):",
                                 vals$gene_of_interest),
                     subtitle="filtered, normalized, variance-stabilized") +
                theme_Publication()
            p 
        } else {
            # Make a heatmap for all the genes using the Log2CPM values
            myheatcolors <- RdBu(75)
            
            diffGenes <- diffGenes.df %>%
                dplyr::select(-geneID) %>%
                as.matrix()
            rownames(diffGenes) <- rownames(v.DEGList.filtered.norm$E)
            colnames(diffGenes) <- as.character(v.DEGList.filtered.norm$targets$group)
            clustColumns <- hclust(as.dist(1-cor(diffGenes, method="spearman")), method="complete")
            subset.diffGenes<- diffGenes[vals$gene_of_interest,]
            clustRows <- hclust(as.dist(1-cor(t(subset.diffGenes), 
                                              method="pearson")), 
                                method="complete") 
            par(cex.main=1.2)
            p<-heatmap.2(subset.diffGenes, 
                         srtCol = 45, adjCol= c(1,1),
                         Rowv= as.dendrogram(clustRows),
                         Colv=as.dendrogram(clustColumns),
                         lmat=rbind( c(0, 3, 4), c(2,1,0 ) ), 
                         lwid=c(0.2, 5, 1 ) ,
                         dendrogram = 'both',
                         key.title = NA,
                         main = paste0("Log2 Counts Per Million (CPM) Expression Across Life Stages"),
                         col=rev(myheatcolors), 
                         scale='row', 
                         labRow=vals$gene_of_interest,
                         density.info="none", 
                         trace="none",
                         cexRow=1.2, cexCol=1.2) 
            p
        }
        
        
        
        
        # p<- ggplot(gene_vals) + 
        #     aes(x = life_stage, y = log2CPM, fill = life_stage) +
        #     geom_boxplot(show.legend = F, alpha = 0.7) +
        #     labs(y="log2 CPM expression", x = "Life Stage",
        #          title=paste("Log2 Counts per Million (CPM):",
        #                      vals$gene_of_interest),
        #          subtitle="filtered, normalized, variance-stabilized") +
        #     #coord_fixed()+
        #     theme_Publication() 
        p
    })
    
    observeEvent(input$goLifeStage_GW,{
        vals$comparison <- NULL
        vals$targetStage <- NULL
        vals$contrastStage <- NULL
        vals$multipleCorrection <- NULL
        vals$results <- NULL
        vals$ebFit <- NULL
    }, ignoreInit = TRUE)
   
    
    ## GW: Pairwise Comparisons Across Life Stage ----
    ### Parse the inputs
    parse_contrasts_GW <- eventReactive(input$goLifeStage_GW,{
        
        if (isTruthy(input$multiContrasts_GW)) {
            comparison <- input$multiContrasts_GW %>%
                gsub(" ", "", ., fixed = TRUE) %>%
                str_split(pattern = ",") %>%
                unlist()
            targetStage<- comparison %>%
                str_split(pattern="-", simplify = T) %>%
                .[,1] %>%
                gsub("(", "", ., fixed = TRUE) %>%
                gsub(")", "", ., fixed = TRUE) %>%
                str_split(pattern = "\\+", simplify = T)
            contrastStage<-comparison %>%
                str_split(pattern="-", simplify = T) %>%
                .[,2] %>%
                gsub("(", "", ., fixed = TRUE) %>%
                gsub(")", "", ., fixed = TRUE) %>%
                str_split(pattern = "\\+", simplify = T)
            if (input$multipleContrastsYN_GW == T) {
                vals$multipleCorrection <- T
            } else vals$multipleCorrection <- F
        } else if (length(input$selectTarget_GW) > 1 | 
                   length(input$selectContrast_GW) > 1) {
            targetStage <- rbind(input$selectTarget_GW)
            contrastStage <- rbind(input$selectContrast_GW)
            comparison <- paste(paste0(input$selectTarget_GW, 
                                       collapse = "+") %>%
                                    paste0("(",.,")"), 
                                paste0(input$selectContrast_GW, 
                                       collapse = "+") %>%
                                    paste0("(",.,")"), 
                                sep = "-")
            vals$multipleCorrection <- F
        } else {
            targetStage <- rbind(input$selectTarget_GW)
            contrastStage <- rbind(input$selectContrast_GW)
            comparison <- paste(input$selectTarget_GW, 
                                input$selectContrast_GW, 
                                sep = "-")
            vals$multipleCorrection <- F
        }
        
        vals$targetStage <- targetStage 
        vals$contrastStage <- contrastStage
        vals$comparison <- comparison
        
    })
    
    ### GW: Generate Responsive Selection for Life Stage to Display ----
    output$contrastDisplaySelection_GW <- renderUI({
        #req(input$goLifeStage_GW) #Remove?
        comparison <- parse_contrasts_GW()
        
        panel(
            heading = tagList(h5(shiny::icon("fas fa-filter"),
                                 "Step 2B: Select Contrast to Display")),
            status = "default",
            selectInput("displayedComparison_GW",
                        NULL, 
                        comparison)
        )
    })
    
    ### GW: Set Contrast Matrix and Fit the Linear Model ----
    set_linear_model_GW <- eventReactive(input$goLifeStage_GW,{
        source('Server/limma_ranking.R', local = TRUE)
    })
    
    ### GW: Extract the Differentially Expressed Genes ----
    pull_DEGs_GW <- reactive({
        req(vals$ebFit)
        if (isTruthy(input$displayedComparison_GW)){
            vals$displayedComparison <- match(input$displayedComparison_GW,
                                              vals$comparison)
        } else {vals$displayedComparison <- 1}
        
        myTopHits.df <- calc_DEG_tbl(vals$ebFit, 
                                     vals$displayedComparison)
        
        #### Filter dataset looking for the genes on the list
        highlight.df <- myTopHits.df %>%
            dplyr::filter(geneID %in% vals$genelist[[1]]) %>%
            dplyr::select(geneID, 
                          logFC, 
                          BH.adj.P.Val:percent_homology)
        vals$highlight.df <- highlight.df
        
        #browser()
        #### Volcano Plots
        vplot <- ggplot(myTopHits.df) +
            aes(y=-log10(BH.adj.P.Val), x=logFC) +
            geom_point(size=2,
                       na.rm = T) +
            geom_point(data = highlight.df, 
                       aes(y=-log10(BH.adj.P.Val), 
                           x=logFC, 
                           color = geneID, 
                           size = 2),
                       na.rm = T) +
            geom_hline(yintercept = -log10(adj.P.thresh), 
                       linetype="longdash", 
                       colour="grey", 
                       size=1) + 
            geom_vline(xintercept = lfc.thresh, 
                       linetype="longdash", 
                       colour="#2C467A", 
                       size=1) +
            geom_vline(xintercept = -lfc.thresh, 
                       linetype="longdash", 
                       colour="#2C467A", 
                       size=1) +
            guides(size = FALSE,
                   colour = guide_legend(override.aes = list(size = 6)))+
            labs(title = paste0('Pairwise Comparison: ',
                                gsub('-',
                                     ' vs ',
                                     vals$comparison[vals$displayedComparison])),
                 color = "GeneIDs") +
            #coord_fixed()+
            theme_Publication()

        vplot
    })
    
    ## GW: Volcano Plot, Generate UI  ----
    output$volcano_GW <- renderPlot({
        req(input$goLifeStage_GW)
        req(vals$genelist)
        set_linear_model_GW()
        pull_DEGs_GW()
    })
    
    output$hover_info <- renderUI({
        req(vals$highlight.df)
        #browser()
        
        pointer.df <- vals$highlight.df %>%
            dplyr::mutate(log10.adj.P.Val = -log10(BH.adj.P.Val))
        
        hover <- input$plot_hover
        point <- nearPoints(pointer.df, hover, 
                            xvar = "logFC",
                            yvar = "log10.adj.P.Val",
                            threshold = 5, maxpoints = 1, addDist = TRUE)
        if (nrow(point) == 0) return(NULL)
        
        
        # calculate point position INSIDE the image as percent of total dimensions
        # from left (horizontal) and from top (vertical)
        left_pct <- (hover$x - hover$domain$left) / (hover$domain$right - hover$domain$left)
        top_pct <- (hover$domain$top - hover$y) / (hover$domain$top - hover$domain$bottom)
        
        # calculate distance from left and bottom side of the picture in pixels
        left_px <- hover$range$left + left_pct * (hover$range$right - hover$range$left)
        top_px <- hover$range$top + top_pct * (hover$range$bottom - hover$range$top)
        
        # create style property fot tooltip
        # background color is set so tooltip is a bit transparent
        # z-index is set so we are sure are tooltip will be on top
        style <- paste0("position:absolute; z-index:100; background-color: rgba(245, 245, 245, 0.85); ",
                        "left:", left_px - 200 , "px; top:", top_px - 200, "px;")
        
        # actual tooltip created as wellPanel
        wellPanel(
            style = style,
            p(HTML(paste0("<b> GeneID: </b>", 
                          point$geneID, 
                          "<br/>", 
                          "<b> Log FC: </b>",
                          round(point$logFC,digits = 2),
                          "<br/>",
                          "<b> p-value: </b>",
                          format(point$BH.adj.P.Val, digits = 3, scientific = TRUE))))
        )
    })
    
    
    #### GW: Data Table of Differentially Expressed Genes from User Subset ----
    assemble_DEGs_GW <- reactive({
        req(vals$displayedComparison)
        tS<- vals$targetStage[vals$displayedComparison,
                              ][vals$targetStage[vals$displayedComparison,
                                                 ]!=""]
        cS<- vals$contrastStage[vals$displayedComparison,
                                ][vals$contrastStage[vals$displayedComparison,
                                                     ]!=""]
        
        # Correct for multiple tests, if necessary
        if (vals$multipleCorrection == T) {
            diffGenes.df <- v.DEGList.filtered.norm$E[vals$results[,
                                                                   vals$displayedComparison] !=0,] %>%
                as_tibble(rownames = "geneID", .name_repair = "unique")
        }
        
        # Get log2CPM values for genes of interest
        highlight.tbl <- diffGenes.df %>%
            dplyr::filter(geneID %in% vals$genelist[[1]]) %>%
            dplyr::select(geneID, starts_with(paste0(tS,"-")), 
                          starts_with(paste0(cS,"-"))) %>%
            left_join(vals$highlight.df, by = "geneID")
        
        highlight.tbl$BH.adj.P.Val <- formatC(highlight.tbl$BH.adj.P.Val, 
                                              digits = 3, 
                                              format = "E") 
        
        
        n_num_cols <- length(tS)*3 + length(cS)*3 + 2
        highlight.datatable <- highlight.tbl %>%
            DT::datatable(extensions = c('KeyTable', "FixedHeader"),
                          rownames = FALSE,
                          caption = htmltools::tags$caption(
                              style = 'caption-side: top; text-align: left; color: black',
                              htmltools::tags$b('Differentially Expressed Genes in', 
                                                htmltools::tags$em('S. stercoralis'), 
                                                gsub('-',' vs ',vals$comparison[vals$displayedComparison])),
                              htmltools::tags$br(),
                              "Threshold: p < ",
                              adj.P.thresh, "; log-fold change > ",
                              lfc.thresh,
                              htmltools::tags$br(),
                              'Values = log2 counts per million'),
                          options = list(keys = TRUE,
                                         autoWidth = TRUE,
                                         scrollX = TRUE,
                                         order = list(n_num_cols-1, 
                                                      'desc'),
                                         searchHighlight = TRUE, 
                                         pageLength = 10, 
                                         lengthMenu = c("10", 
                                                        "25", 
                                                        "50", 
                                                        "100"),
                                         columnDefs = list(
                                             list(
                                                 targets = ((n_num_cols + 
                                                                 4):(n_num_cols + 
                                                                         5)),
                                                 render = JS(
                                                     "function(data, type, row, meta) {",
                                                     "return type === 'display' && data.length > 20 ?",
                                                     "'<span title=\"' + data + '\">' + data.substr(0, 20) + '...</span>' : data;",
                                                     "}")
                                             ),
                                             list(targets = "_all",
                                                  class="dt-right")
                                         )
                                         
                          )) 
        highlight.datatable <- highlight.datatable %>%
            DT::formatRound(columns=c(2:n_num_cols,
                                      n_num_cols+2, 
                                      n_num_cols+8), 
                            digits=2)
        
        highlight.datatable
    })
    
    
    output$highlight.df <- renderDT ({
        req(vals$highlight.df)
        assemble_DEGs_GW()
    })
    
    
    
    ## LS: Display pairwise comparisons ----
    ### Parse the inputs
    ### 
    observeEvent(input$goLS,{
        vals$comparison <- NULL
        vals$targetStage <- NULL
        vals$contrastStage <- NULL
        vals$multipleCorrection <- NULL
        vals$results <- NULL
        vals$ebFit <- NULL
        updateSelectInput(session, "displayedComparison_LS", selected = "", choices = NULL)
    }, ignoreInit = TRUE)
    
    parse_contrasts_LS<-eventReactive(input$goLS,{
        if (isTruthy(input$multiContrasts_LS)) {
            comparison <- input$multiContrasts_LS %>%
                gsub(" ", "", ., fixed = TRUE) %>%
                str_split(pattern = ",") %>%
                unlist()
            targetStage<- comparison %>%
                str_split(pattern="-", simplify = T) %>%
                .[,1] %>%
                gsub("(", "", ., fixed = TRUE) %>%
                gsub(")", "", ., fixed = TRUE) %>%
                str_split(pattern = "\\+", simplify = T)
            contrastStage<-comparison %>%
                str_split(pattern="-", simplify = T) %>%
                .[,2] %>%
                gsub("(", "", ., fixed = TRUE) %>%
                gsub(")", "", ., fixed = TRUE) %>%
                str_split(pattern = "\\+", simplify = T)
            if (input$multipleContrastsYN_LS == T) {
                vals$multipleCorrection <- T
            } else vals$multipleCorrection <- F
        } else if (length(input$selectTarget_LS > 1) | 
                   length(input$selectContrast_LS > 1)){
            targetStage <- rbind(input$selectTarget_LS)
            contrastStage <- rbind(input$selectContrast_LS)
            comparison <- paste(paste0(input$selectTarget_LS, 
                                       collapse = "+") %>%
                                    paste0("(",.,")"), 
                                paste0(input$selectContrast_LS, 
                                       collapse = "+") %>%
                                    paste0("(",.,")"), 
                                sep = "-")
            vals$multipleCorrection <- F
        } else {
            targetStage <- rbind(input$selectTarget_LS)
            contrastStage <- rbind(input$selectContrast_LS)
            comparison <- paste(input$selectTarget_LS, 
                                input$selectContrast_LS, 
                                sep = "-")
            vals$multipleCorrection <- F
        }
        
        vals$targetStage <- targetStage 
        vals$contrastStage <- contrastStage
        vals$comparison <- comparison
        
    })
    
    ### LS: Generate Responsive Selection for Life Stage to Display ----
    output$contrastDisplaySelection_LS <- renderUI({
        comparison <- parse_contrasts_LS()
        panel(
            heading = tagList(h5(shiny::icon("fas fa-filter"),
                                 "Step 1B: Select Contrast to Display")),
            status = "default",
            selectInput("displayedComparison_LS",
                        "Choose a Comparison to Display", 
                        comparison)
        )
    })
    
    ### LS: Set Contrast Matrix and Fit the Linear Model ----
    set_linear_model_LS <- eventReactive(input$goLS,{
        source('Server/limma_ranking.R', local = TRUE)
    })
    
    ### LS: Extract the Differentially Expressed Genes ----
    pull_DEGs_LS <- reactive({
        req(vals$ebFit)
        if (isTruthy(input$displayedComparison_LS)){
            vals$displayedComparison <- match(input$displayedComparison_LS, 
                                              vals$comparison)
        } else {vals$displayedComparison <- 1}
        
        myTopHits.df <- calc_DEG_tbl(vals$ebFit, 
                                     vals$displayedComparison)
        
        #### Filter dataset for desired columns to display
        highlight.df <- myTopHits.df %>%
            dplyr::select(geneID, 
                          logFC, 
                          BH.adj.P.Val:percent_homology)
        vals$highlight.df <- highlight.df
        
        #### Volcano Plots
        vplot <- ggplot(myTopHits.df) +
            aes(y=-log10(BH.adj.P.Val), x=logFC) +
            geom_point(size=2) +
            geom_hline(yintercept = -log10(adj.P.thresh), 
                       linetype="longdash", 
                       colour="grey", 
                       size=1) + 
            geom_vline(xintercept = lfc.thresh, 
                       linetype="longdash", 
                       colour="#BE684D", 
                       size=1) +
            geom_vline(xintercept = -lfc.thresh, 
                       linetype="longdash", 
                       colour="#2C467A", 
                       size=1) +
            labs(title = paste0('Pairwise Comparison: ',
                                gsub('-',
                                     ' vs ',
                                     vals$comparison[vals$displayedComparison])),
                 color = "GeneIDs") +
            theme_Publication() 
        vplot
    })
    
    ### LS: Volcano Plot, Generate UI ----
    output$volcano_LS <- renderPlot({
        req(input$goLS)
        set_linear_model_LS()
        pull_DEGs_LS()
    })
    
    output$hover_info_LS <- renderUI({
        req(vals$highlight.df)
        
        pointer.df <- vals$highlight.df %>%
            dplyr::mutate(log10.adj.P.Val = -log10(BH.adj.P.Val))
        
        hover <- input$plot_hover_LS
        point <- nearPoints(pointer.df, hover, 
                            xvar = "logFC",
                            yvar = "log10.adj.P.Val",
                            threshold = 5, maxpoints = 1, addDist = TRUE)
        if (nrow(point) == 0) return(NULL)
        
        
        # calculate point position INSIDE the image as percent of total dimensions
        # from left (horizontal) and from top (vertical)
        left_pct <- (hover$x - hover$domain$left) / (hover$domain$right - hover$domain$left)
        top_pct <- (hover$domain$top - hover$y) / (hover$domain$top - hover$domain$bottom)
        
        # calculate distance from left and bottom side of the picture in pixels
        left_px <- hover$range$left + left_pct * (hover$range$right - hover$range$left)
        top_px <- hover$range$top + top_pct * (hover$range$bottom - hover$range$top)
        
        # create style property fot tooltip
        # background color is set so tooltip is a bit transparent
        # z-index is set so we are sure are tooltip will be on top
        style <- paste0("position:absolute; z-index:100; background-color: rgba(245, 245, 245, 0.85); ",
                        "left:", left_px - 200 , "px; top:", top_px - 200, "px;")
        
        # actual tooltip created as wellPanel
        wellPanel(
            style = style,
            p(HTML(paste0("<b> GeneID: </b>", 
                          point$geneID, 
                          "<br/>", 
                          "<b> Log FC: </b>",
                          round(point$logFC,digits = 2),
                          "<br/>",
                          "<b> p-value: </b>",
                          format(point$BH.adj.P.Val, digits = 3, scientific = TRUE))))
        )
    })
    
    #### LS: Data Table of Differentially Expressed Genes ----
    assemble_DEGs_LS <- reactive({
        req(vals$displayedComparison)
        tS<- vals$targetStage[vals$displayedComparison,
                              ][vals$targetStage[vals$displayedComparison,
                                                 ]!=""]
        cS<- vals$contrastStage[vals$displayedComparison,
                                ][vals$contrastStage[vals$displayedComparison,
                                                     ]!=""]
        
        # Correct for multiple tests, if necessary
        if (vals$multipleCorrection == T) {
            diffGenes.df <- v.DEGList.filtered.norm$E[vals$results[,
                                                                   vals$displayedComparison] !=0,] %>%
                as_tibble(rownames = "geneID", .name_repair = "unique")
        }
        
        # Get log2CPM values for life stages of interest
        LS.tbl <- diffGenes.df %>%
            dplyr::select(geneID, starts_with(paste0(tS,"-")), 
                          starts_with(paste0(cS,"-"))) %>%
            left_join(vals$highlight.df, by = "geneID")
        
        
        LS.tbl$BH.adj.P.Val <- formatC(LS.tbl$BH.adj.P.Val, 
                                       digits = 3, 
                                       format = "E") 
        
        n_num_cols <- length(tS)*3 + length(cS)*3 + 2
        LS.datatable <- LS.tbl %>%
            DT::datatable(extensions = c('KeyTable', "FixedHeader"),
                          rownames = FALSE,
                          caption = htmltools::tags$caption(
                              style = 'caption-side: top; text-align: left; color: black;',
                              htmltools::tags$b('Differentially Expressed Genes in', 
                                                htmltools::tags$em('S. stercoralis'), 
                                                gsub('-',' vs ',vals$comparison[vals$displayedComparison])),
                              htmltools::tags$br(),
                              "Threshold: p < ",
                              adj.P.thresh, "; log-fold change > ",
                              lfc.thresh,
                              htmltools::tags$br(),
                              'Values = log2 counts per million'),
                          options = list(keys = TRUE,
                                         autoWidth = TRUE,
                                         scrollX = TRUE,
                                         order = list(n_num_cols-1, 
                                                      'desc'),
                                         searchHighlight = TRUE, 
                                         pageLength = 10, 
                                         lengthMenu = c("10", 
                                                        "25", 
                                                        "50", 
                                                        "100"),
                                         columnDefs = list(
                                             list(
                                                 targets = ((n_num_cols + 
                                                                 4):(n_num_cols + 
                                                                         5)),
                                                 render = JS(
                                                     "function(data, type, row, meta) {",
                                                     "return type === 'display' && data.length > 20 ?",
                                                     "'<span title=\"' + data + '\">' + data.substr(0, 20) + '...</span>' : data;",
                                                     "}")
                                             ),
                                             list(targets = "_all",
                                                  class="dt-right")
                                         )
                                         
                          )) 
        LS.datatable <- LS.datatable %>%
            DT::formatRound(columns=c(2:n_num_cols,
                                      n_num_cols+2, 
                                      n_num_cols+8), 
                            digits=2)
        
        LS.datatable
    })
    
    
    output$tbl_LS <- renderDT ({
        req(input$goLS)
        DEG.datatable_LS<-assemble_DEGs_LS()
        DEG.datatable_LS
        
    })
    
    session$onSessionEnded(stopApp)
}

# Run the application 
shinyApp(ui = ui, server = server)
