# Stercoralis RNAseq Browser Shiny App

## ---- Libraries ----
suppressPackageStartupMessages({
    library(shiny)
    library(shinyWidgets)
    library(htmltools)
    library(shinythemes)
    library(DT)
    library(tidyverse)
    library(limma) # differential gene expression using linear modeling
    library(edgeR)
    library(gt) 
    library(plotly)
    library(shinycssloaders)
    library(magrittr)
    library(ggthemes)
    library(gplots)
    library(svglite)
    library(Cairo)
    library(heatmaply)
    library(RColorBrewer)
    library(openxlsx)
    library(egg)
    library(dendextend)
    source("Server/ggheatmap_local.R")
    source("Server/calc_DEG_tbl.R")
    source("Server/theme_Publication.R")
    source('Server/limma_ranking.R')
    source("Server/excel_srv.R")
})

## ---- Background ----
# Load pre-calculated data generated by S_stercoralis_RNAseq_ReAnalysis.R
# 
# Import a variance-stabilized DEGList created by voom transformation command.
# Outputs: E = normalized CPMexpression values on the log2 scale
load(file = "./vDEGList")

# Parse vDEGList into a tibble containing Log2CPM information
Log2CPM<-v.DEGList.filtered.norm$E %>%
    as_tibble(rownames = "geneID")%>%
    setNames(nm = c("geneID", 
                    as.character(v.DEGList.filtered.norm$targets$group))) %>%
    pivot_longer(cols = -geneID,
                 names_to = "life_stage", 
                 values_to = "log2CPM") %>%
    group_by(geneID, life_stage)

diffGenes.df <- v.DEGList.filtered.norm$E %>%
    as_tibble(rownames = "geneID", .name_repair = "unique")

# Fit a linear model to the data ----
fit <- lmFit(v.DEGList.filtered.norm, v.DEGList.filtered.norm$design)


# Set Expression threshold values for plotting and saving DEGs ----
adj.P.thresh <- 0.05
lfc.thresh <- 1

## ---- UI ----

ui <- fluidPage(
    
    #source('UI/SsRNAseqBrowser_dashboard-ui.R', local = T)$value
    source('UI/SsRNAseqBrowser_navbar-ui.R', local = T)$value,
    
    # Custom CSS ----
    tags$head(
        tags$style(HTML("
    h3 {
    font-size: 22px;
    margin: 10.5px 0px;
    }
    
    .navbar-brand {
    height: 80px;
    }
    
    .navbar-nav>li>a{
    height: 80px;
    }
    
    h4 {
    font-size: 16px;
    font-weight: bold;
    }
    
    h5 {
    font-size: 15px;
    font-weight: bold;
    }
    
    h6 {
    font-size: 13px;
    margin: 0px;
    font-weight: 550;
    line-height: 1.4;
    }
    
    strong {
    font-size: 13px;
    font-weight: bold;
    }
    
    p{
    font-size: 13px;
    font-weight: normal;
    }
    
    .selectize-input {
    word-wrap: break-word;
    font-size: 13px;
    overflow-x: auto;
    }
    
    .selectize-dropdown {
    word-wrap: break-word;
    font-size: 13px;
    }
    
    .form-control {
    font-size: 13px;
    height: 40px;
    }
    
    .btn {
    font-size: 13px;
    height: 40px;
    }
    
    
                  
                    "))
        
    )
    
)


# Define server logic  ----
server <- function(input, output, session) {
    vals<-reactiveValues()
    
    ## GW: Generate/Reset Gene File Upload ----
    output$genefile_upload <- renderUI({
        input$resetGenes
        fileInput('loadfile',
                  h6('Gene Stable ID List (.csv)'),
                  multiple = FALSE)
    })
    
    ## GW: Clear Genes ----
    observeEvent(input$resetGenes,{
        updateTextAreaInput(session,"idtext",value = "")
        vals$genelist <- NULL
    })
    
    ## GW: Parse Gene Inputs ----
    parse_ids <- eventReactive(input$goGW,{
        isolate({
            if (isTruthy(input$idtext)){
                if (any(grepl('SSTP', input$idtext, ignore.case = TRUE))){
                    # Text input matches SSTP values
                    genelist <- input$idtext %>%
                        gsub(" ", "", ., fixed = TRUE) %>%
                        str_split(pattern = ",") %>%
                        unlist() %>%
                        as_tibble_col(column_name = "geneID")
                } else {
                    # Assume the values here are keywords, search Description terms
                    terms <- input$idtext %>%
                        str_split(pattern = ",") %>%
                        unlist()
                    geneindex<-sapply(terms, function(y) {
                        grepl(gsub("$ |^ ","",y), 
                              v.DEGList.filtered.norm$genes$Description,
                              ignore.case = TRUE)
                    }) %>%
                        rowSums() %>%
                        as.logical()
                    genelist <- v.DEGList.filtered.norm$genes %>%
                        rownames_to_column(var = "geneID") %>%
                        dplyr::select(geneID)
                    genelist <- genelist$geneID[geneindex] %>%
                        as_tibble_col(column_name = "geneID")
                }
                
            } else if (isTruthy(input$loadfile)){
                file <- input$loadfile
                ext <- tools::file_ext(file$datapath)
                validate(need(ext == "csv", "Please upload a csv file"))
                genelist <- read.csv(file$datapath, 
                                     header = FALSE, 
                                     colClasses = "character", 
                                     strip.white = T) %>%
                    as_tibble() %>%
                    pivot_longer(cols = everything(), 
                                 values_to = "geneID") %>%
                    dplyr::select(geneID)
            } 
            
            vals$genelist <- genelist
            vals$genelist.Log2CPM <- Log2CPM %>%
                dplyr::filter(geneID %in% genelist$geneID)
        })
    })
    
    ### GW: Generate Responsive Selection for Gene to Display ----
    output$geneDisplaySelection_GW <- renderUI({
        parse_ids()
        req(vals$genelist)
        isolate({
            if (length(vals$genelist$geneID)>1) {
                choices <- c("All Genes", vals$genelist$geneID)
                selected <- "All Genes"
            } else choices <- vals$genelist$geneID
            panel(
                heading = tagList(h4(shiny::icon("fas fa-filter"),
                                     "Pick Gene to Display")),
                status = "default",
                pickerInput("displayedGene",
                            NULL, 
                            choices,
                            options = list(style = 'btn btn-primary'))
            )
        })
    })
    
    ## GW: Plots of Gene Expression ----
    generateGenePlot <- reactive({
        req(vals$genelist)
        req(input$displayedGene)
        req(vals$genelist.Log2CPM)
        # Select gene to display
        if (isTruthy(input$displayedGene)){
            if (input$displayedGene == "All Genes"){
                vals$gene_of_interest <- vals$genelist$geneID
            } else vals$gene_of_interest <- input$displayedGene
        } else {vals$gene_of_interest <- vals$genelist[[1,1]]}
        
        
        if (length(vals$gene_of_interest) == 1) {
            # Plot Log2CPM values for an individual gene
            gene_vals <- vals$genelist.Log2CPM %>%
                dplyr::filter(geneID == vals$gene_of_interest)
            
            p<- ggplot(gene_vals) + 
                aes(x = life_stage, y = log2CPM, fill = life_stage) +
                geom_boxplot(show.legend = F, alpha = 0.7) +
                labs(y="log2 CPM expression", x = "Life Stage",
                     title=paste("Log2 Counts per Million (CPM):",
                                 vals$gene_of_interest),
                     subtitle="filtered, normalized, variance-stabilized") +
                theme_Publication() + 
                theme(aspect.ratio=2/3) 
            p 
        } else {
            # Make a heatmap for all the genes using the Log2CPM values
            myheatcolors <- RdBu(75)
            
            diffGenes <- diffGenes.df %>%
                dplyr::select(-geneID) %>%
                as.matrix()
            rownames(diffGenes) <- rownames(v.DEGList.filtered.norm$E)
            clustColumns <- hclust(as.dist(1-cor(diffGenes, method="spearman")), method="complete")
            subset.diffGenes<- diffGenes[vals$gene_of_interest,]
            colnames(subset.diffGenes) <- paste0(v.DEGList.filtered.norm$targets$group, 
                                                 "_", 
                                                 rep(1:3,7))
            clustRows <- hclust(as.dist(1-cor(t(subset.diffGenes), 
                                              method="pearson")), 
                                method="complete") 
            par(cex.main=1.2)
            
            hovertext <- as.data.frame(subset.diffGenes) %>%
                round(digits = 2)
            
            hovertext[] <- lapply(hovertext, function(x){
                paste0("Log2CPM: ", x)
            })
            
            showticklabels <- if(length(vals$gene_of_interest)<20){c(TRUE,TRUE)} else {c(TRUE,FALSE)}
            
            p <- heatmaply(subset.diffGenes,
                           colors = rev(myheatcolors),
                           Rowv= ladderize(as.dendrogram(clustRows)),
                           Colv=ladderize(as.dendrogram(clustColumns)),
                           show_dendrogram = c(FALSE, TRUE),
                           showticklabels = showticklabels,
                           scale='row', #rows are scaled to have mean zero and standard deviation one. 
                           plot_method = "ggplot",
                           branches_lwd = 0.2,
                           key.title = "Z-Score",
                           cexRow=1.2, cexCol=1.2,
                           main = paste0("Log2 Counts Per Million (CPM) Expression Across Life Stages"),
                           custom_hovertext = hovertext)
        }
    })
    
    
    
    
    ### GW: Switch what type of GW Plot is produced ----
    output$CPM <- renderPlot({
        req(input$displayedGene != "All Genes")
        generateGenePlot()
    })
    
    output$CPMPlotly <- renderPlotly({
        req(input$displayedGene == "All Genes")
        generateGenePlot()
    })
    
    observeEvent(input$displayedGene,{
        if(input$displayedGene == "All Genes"){
            removeUI(
                selector = "#CPMPlotlydiv"
            )
            insertUI(
                selector = '#test',
                where = "afterBegin",
                ui = tagList(div(id = "CPMPlotlydiv",
                                 withSpinner(plotlyOutput('CPMPlotly'),
                                             color = "#2C3E50")
                ))
            )
            removeUI(
                selector = "#CPMdiv"
            )
            
        }else{
            removeUI(
                selector = "#CPMdiv"
            )
            insertUI(
                selector = '#test',
                where = "afterBegin",
                ui = tagList(div(id = "CPMdiv",
                                 withSpinner(plotOutput('CPM'),
                                             color = "#2C3E50")
                                 ))
            )
            removeUI(
                selector = "#CPMPlotlydiv"
            )
        }
    })
    
    ## GW: Save Gene Plots ----
    output$downloadGenePlot <- downloadHandler(
        filename = function(){
            paste(input$displayedGene, '_',Sys.Date(),'.pdf', sep='')
        },
        content = function(file){
           
            
            if (input$displayedGene == "All Genes") {
                
                # Make a heatmap for all the genes using the Log2CPM values
                myheatcolors <- RdBu(75)
                
                diffGenes <- diffGenes.df %>%
                    dplyr::select(-geneID) %>%
                    as.matrix()
                rownames(diffGenes) <- rownames(v.DEGList.filtered.norm$E)
                colnames(diffGenes) <- as.character(v.DEGList.filtered.norm$targets$group)
                clustColumns <- hclust(as.dist(1-cor(diffGenes, method="spearman")), method="complete")
                subset.diffGenes<- diffGenes[vals$gene_of_interest,]
                colnames(subset.diffGenes) <- paste0(v.DEGList.filtered.norm$targets$group, 
                                                     "_", 
                                                     rep(1:3,7))
                clustRows <- hclust(as.dist(1-cor(t(subset.diffGenes), 
                                                  method="pearson")), 
                                    method="complete") 
                par(cex.main=1.2)
                
                showticklabels <- if(length(vals$gene_of_interest)<20){c(TRUE,TRUE)} else {c(TRUE,FALSE)}
                p<-ggheatmap_local(subset.diffGenes,
                                   colors = rev(myheatcolors),
                                   Rowv= ladderize(as.dendrogram(clustRows)),
                                   Colv=ladderize(as.dendrogram(clustColumns)),
                                   show_dendrogram = c(FALSE, TRUE),
                                   key.title = "Z-Score",
                                   branches_lwd = 0.5,
                                   showticklabels = showticklabels,
                                   scale='row',
                                   cexRow=1.2, cexCol=1.2,
                                   #margin = c(50,1000),
                                   main = "Log2 Counts Per Million (CPM) Expression Across Life Stages")
                ggsave(file, 
                       plot = p,
                       width = 11,
                       height = 5, 
                       device = "pdf", 
                       useDingbats=FALSE)
            } else {
                p<-generateGenePlot()
                
                ggsave(file, 
                       plot = p,
                       width = 7,
                       height = 7, 
                       device = "pdf", 
                       useDingbats=FALSE)
            }
        }
        
    )
    
    ## GW: Clear Comparison Selections ----
    observeEvent(input$resetGW,{
        updateSelectInput(session, "selectContrast_GW", selected = "")
        updateSelectInput(session, "selectTarget_GW", selected = "")
        updateTextAreaInput(session,"multiContrasts_GW",value = "")
        
    })
    
    ## GW: Pairwise Comparisons Across Life Stage ----
    
    ### Parse the inputs
    parse_contrasts_GW <- eventReactive(input$goLifeStage_GW,{
        if (isTruthy(input$multiContrasts_GW)) {
            comparison <- input$multiContrasts_GW %>%
                gsub(" ", "", ., fixed = TRUE) %>%
                str_split(pattern = "[,;]") %>%
                unlist()
            targetStage<- comparison %>%
                str_split(pattern="-", simplify = T) %>%
                .[,1] %>%
                gsub("(", "", ., fixed = TRUE) %>%
                gsub(")", "", ., fixed = TRUE) %>%
                str_split(pattern = "\\+", simplify = T)
            
            contrastStage<-comparison %>%
                str_split(pattern="-", simplify = T) %>%
                .[,2] %>%
                gsub("(", "", ., fixed = TRUE) %>%
                gsub(")", "", ., fixed = TRUE)  %>%
                str_split(pattern = "\\+", simplify = T)
            
            comparison<- sapply(seq_along(comparison),function(x){
                tS <- as.vector(targetStage[x,]) %>%
                    .[. != ""] 
                cS <- as.vector(contrastStage[x,]) %>%
                    .[. != ""] 
                paste(paste0(tS, 
                             collapse = "+") %>%
                          paste0("(",.,")/",length(tS)),
                      paste0(cS, 
                             collapse = "+") %>%
                          paste0("(",.,")/",length(cS)),
                      sep = "-")
                
            })
            if (input$multipleContrastsYN_GW == T) {
                vals$multipleCorrection_GW <- T
            } else vals$multipleCorrection_GW <- F
        } else if (length(input$selectTarget_GW) > 1 | 
                   length(input$selectContrast_GW) > 1) {
            targetStage <- rbind(input$selectTarget_GW)
            contrastStage <- rbind(input$selectContrast_GW)
            comparison <- paste(paste0(input$selectTarget_GW, 
                                       collapse = "+") %>%
                                    paste0("(",.,")/",length(input$selectTarget_GW)), 
                                paste0(input$selectContrast_GW, 
                                       collapse = "+") %>%
                                    paste0("(",.,")/",length(input$selectContrast_GW)), 
                                sep = "-")
            vals$multipleCorrection_GW <- F
        } else {
            targetStage <- rbind(input$selectTarget_GW)
            contrastStage <- rbind(input$selectContrast_GW)
            comparison <- paste(input$selectTarget_GW, 
                                input$selectContrast_GW, 
                                sep = "-")
            vals$multipleCorrection_GW <- F
        }
        
        vals$targetStage_GW <- targetStage 
        vals$contrastStage_GW <- contrastStage
        vals$limmacontrast_GW <- comparison
        vals$comparison_GW <- gsub("/[0-9]*","", comparison)
        
    })
    
    ### GW: Generate Responsive Selection for Life Stage to Display ----
    output$contrastDisplaySelection_GW <- renderUI({
        comparison <- parse_contrasts_GW()
        panel(
            heading = tagList(h4(shiny::icon("fas fa-filter"),
                                 "Pick Contrast to Display")),
            status = "default",
            pickerInput("displayedComparison_GW",
                        NULL, 
                        comparison,
                        selected = comparison[[1]],
                        options = list(style = 'btn btn-primary'))
        )
    })
    
    ### GW: Set Contrast Matrix and Fit the Linear Model ----
    set_linear_model_GW <- eventReactive(input$goLifeStage_GW,{
        req(vals$limmacontrast_GW)
        limma_ranking(vals$limmacontrast_GW, 
                      vals$targetStage_GW, 
                      vals$contrastStage_GW, 
                      vals$multipleCorrection_GW, 
                      vals$genelist, 
                      vals, fit, 
                      v.DEGList.filtered.norm, 
                      adj.P.thresh, 
                      diffGenes.df)
        #vals$ebFit_GW <- vals$ebFit
        vals$list.myTopHits.df_GW <- vals$list.myTopHits.df
        vals$list.highlight.tbl_GW <- vals$list.highlight.tbl
        
    })
    
    ### GW: Extract the Differentially Expressed Genes ----
    pull_DEGs_GW <- reactive({
        
        req(vals$comparison_GW)
        req(vals$list.myTopHits.df_GW)
        if (isTruthy(input$displayedComparison_GW)){
            vals$displayedComparison_GW <- match(input$displayedComparison_GW,
                                                 vals$comparison_GW, nomatch = 1)
        } else {vals$displayedComparison_GW <- 1}
        
        
        #### Volcano Plots
        point_labels <- if(nrow(vals$list.highlight.tbl_GW[[vals$displayedComparison_GW]])<
                           20){guide_legend(override.aes = list(size = 4))} else {FALSE}
        
        vplot <- ggplot(vals$list.myTopHits.df_GW[[vals$displayedComparison_GW]]) +
            aes(y=-log10(BH.adj.P.Val), x=logFC) +
            geom_point(size=3,
                       na.rm = T) +
            geom_point(data = vals$list.highlight.tbl_GW[[vals$displayedComparison_GW]], 
                       mapping = aes(y=-log10(BH.adj.P.Val), 
                                     x=logFC, 
                                     color = geneID),
                       size = 3,
                       na.rm = T) +
            geom_hline(yintercept = -log10(adj.P.thresh), 
                       linetype="longdash", 
                       colour="grey", 
                       size=1) + 
            geom_vline(xintercept = lfc.thresh, 
                       linetype="longdash", 
                       colour="#BE684D", 
                       size=1) +
            geom_vline(xintercept = -lfc.thresh, 
                       linetype="longdash", 
                       colour="#2C467A", 
                       size=1) +
            guides(size = FALSE,
                   colour = point_labels)+
            labs(title = paste0('Pairwise Comparison: ',
                                gsub('-',
                                     ' vs ',
                                     vals$comparison_GW[vals$displayedComparison_GW])),
                 subtitle = paste0("grey line: p = ",
                                   adj.P.thresh, "; colored lines: log-fold change = ",  lfc.thresh),
                 color = "GeneIDs") +
            theme_Publication() +
            theme(aspect.ratio=1/3)
        
        vplot
        
    })
    
    ## GW: Volcano Plot, Generate UI  ----
    output$volcano_GW <- renderPlot({
        req(input$goLifeStage_GW)
        req(vals$comparison_GW)
        
        set_linear_model_GW()
        pull_DEGs_GW()
        
    })
    
    ## GW: Save Volcano Plot ----
    output$downloadVolcano_GW <- downloadHandler(
        filename = function(){
            paste(vals$comparison_GW[vals$displayedComparison_GW], '_',Sys.Date(),'.pdf', sep='')
        },
        content = function(file){
            pull_DEGs_GW()
            ggsave(file,width = 11, height = 8, units = "in", device = "pdf", useDingbats=FALSE)
        }
    )
    
    output$hover_info <- renderUI({
        req(vals$displayedComparison_GW)
        
        pointer.df <- vals$list.highlight.tbl_GW[[vals$displayedComparison_GW]] %>%
            dplyr::mutate(log10.adj.P.Val = -log10(BH.adj.P.Val))
        
        hover <- input$plot_hover
        point <- nearPoints(pointer.df, hover, 
                            xvar = "logFC",
                            yvar = "log10.adj.P.Val",
                            threshold = 5, maxpoints = 1, addDist = TRUE)
        if (nrow(point) == 0) return(NULL)
        
        
        # calculate point position INSIDE the image as percent of total dimensions
        # from left (horizontal) and from top (vertical)
        left_pct <- (hover$x - hover$domain$left) / (hover$domain$right - hover$domain$left)
        top_pct <- (hover$domain$top - hover$y) / (hover$domain$top - hover$domain$bottom)
        
        # calculate distance from left and bottom side of the picture in pixels
        left_px <- hover$range$left + left_pct * (hover$range$right - hover$range$left)
        top_px <- hover$range$top + top_pct * (hover$range$bottom - hover$range$top)
        
        # create style property fot tooltip
        # background color is set so tooltip is a bit transparent
        # z-index is set so we are sure are tooltip will be on top
        style <- paste0("position:absolute; z-index:100; background-color: rgba(245, 245, 245, 0.85); ",
                        "left:", left_px - 200 , "px; top:", top_px - 200, "px;")
        
        # actual tooltip created as wellPanel
        wellPanel(
            style = style,
            p(HTML(paste0("<b> GeneID: </b>", 
                          point$geneID, 
                          "<br/>", 
                          "<b> Log FC: </b>",
                          round(point$logFC,digits = 2),
                          "<br/>",
                          "<b> p-value: </b>",
                          format(point$BH.adj.P.Val, digits = 3, scientific = TRUE))))
        )
    })
    
    
    #### GW: Data Table of Differentially Expressed Genes from User Subset ----
    assemble_DEGs_GW <- reactive({
        req(vals$displayedComparison_GW)
        tS<- vals$targetStage_GW[vals$displayedComparison_GW,
                                 ][vals$targetStage_GW[vals$displayedComparison_GW,
                                                       ]!=""]
        cS<- vals$contrastStage_GW[vals$displayedComparison_GW,
                                   ][vals$contrastStage_GW[vals$displayedComparison_GW,
                                                           ]!=""]
        
        n_num_cols <- length(tS)*3 + length(cS)*3 + 5
        highlight.datatable <- vals$list.highlight.tbl_GW[[vals$displayedComparison_GW]] %>%
            DT::datatable(rownames = FALSE,
                          caption = htmltools::tags$caption(
                              style = 'caption-side: top; text-align: left; color: black',
                              htmltools::tags$b('Differentially Expressed Genes in', 
                                                htmltools::tags$em('S. stercoralis'), 
                                                gsub('-',' vs ',vals$comparison_GW[vals$displayedComparison_GW])),
                              htmltools::tags$br(),
                              "Threshold: p < ",
                              adj.P.thresh, "; log-fold change > ",
                              lfc.thresh,
                              htmltools::tags$br(),
                              'Values = log2 counts per million'),
                          options = list(autoWidth = TRUE,
                                         scrollX = TRUE,
                                         scrollY = '300px',
                                         scrollCollapse = TRUE,
                                         order = list(n_num_cols-1, 
                                                      'desc'),
                                         searchHighlight = TRUE, 
                                         pageLength = 25, 
                                         lengthMenu = c("5",
                                                        "10",
                                                        "25",
                                                        "50",
                                                        "100"),
                                         columnDefs = list(
                                             list(
                                                 targets = ((n_num_cols + 
                                                                 4):(n_num_cols + 
                                                                         5)),
                                                 render = JS(
                                                     "function(data, type, row, meta) {",
                                                     "return type === 'display' && data.length > 20 ?",
                                                     "'<span title=\"' + data + '\">' + data.substr(0, 20) + '...</span>' : data;",
                                                     "}")
                                             ),
                                             list(targets = "_all",
                                                  class="dt-right")
                                         )
                                         
                          )) 
        
        highlight.datatable <- highlight.datatable %>%
            DT::formatRound(columns=c(3:n_num_cols), 
                            digits=3)
        
        highlight.datatable <- highlight.datatable %>%
            DT::formatRound(columns=c(n_num_cols+2, 
                                      n_num_cols+8), 
                            digits=2)
        
        highlight.datatable <- highlight.datatable %>%
            DT::formatSignif(columns=c(n_num_cols+1), 
                             digits=3)
        
        highlight.datatable
    })
    
    
    output$highlight.df <- renderDT ({
        req(vals$list.highlight.tbl_GW)
        req(vals$displayedComparison_GW)
        assemble_DEGs_GW()
    })
    
    # GW: Save Excel Tables with DEG Tables ----
    output$downloadbuttonGW <- renderUI({
        req(input$goLifeStage_GW)
        req(vals$comparison_GW)
        
        output$generate_excel_report_GW <- generate_excel_report(vals$comparison_GW, 
                                                                 vals$list.highlight.tbl_GW)
        downloadButton("generate_excel_report_GW",
                       "Download DEG Tables",
                       class = "btn-primary")
    })
    
    ## LS: Clear Comparison Selections ----
    observeEvent(input$resetLS,{
        updateSelectInput(session, "selectContrast_LS", selected = "")
        updateSelectInput(session, "selectTarget_LS", selected = "")
        updateTextAreaInput(session,"multiContrasts_LS",value = "")
        
    })
    
    ## LS: Pairwise comparisons Across Life Stages ----
    parse_contrasts_LS<-eventReactive(input$goLS,{
        if (isTruthy(input$multiContrasts_LS)) {
            comparison <- input$multiContrasts_LS %>%
                gsub(" ", "", ., fixed = TRUE) %>%
                str_split(pattern = "[,;]") %>%
                unlist()
            targetStage<- comparison %>%
                str_split(pattern="-", simplify = T) %>%
                .[,1] %>%
                gsub("(", "", ., fixed = TRUE) %>%
                gsub(")", "", ., fixed = TRUE) %>%
                str_split(pattern = "\\+", simplify = T)
            contrastStage<-comparison %>%
                str_split(pattern="-", simplify = T) %>%
                .[,2] %>%
                gsub("(", "", ., fixed = TRUE) %>%
                gsub(")", "", ., fixed = TRUE) %>%
                str_split(pattern = "\\+", simplify = T)
            comparison<- sapply(seq_along(comparison),function(x){
                tS <- as.vector(targetStage[x,]) %>%
                    .[. != ""] 
                cS <- as.vector(contrastStage[x,]) %>%
                    .[. != ""] 
                paste(paste0(tS, 
                             collapse = "+") %>%
                          paste0("(",.,")/",length(tS)),
                      paste0(cS, 
                             collapse = "+") %>%
                          paste0("(",.,")/",length(cS)),
                      sep = "-")
                
            })
            if (input$multipleContrastsYN_LS == T) {
                vals$multipleCorrection_LS <- T
            } else vals$multipleCorrection_LS <- F
        } else if (length(input$selectTarget_LS) > 1 | 
                   length(input$selectContrast_LS) > 1){
            targetStage <- rbind(input$selectTarget_LS)
            contrastStage <- rbind(input$selectContrast_LS)
            comparison <- paste(paste0(input$selectTarget_LS, 
                                       collapse = "+") %>%
                                    paste0("(",.,")/",length(input$selectTarget_LS)), 
                                paste0(input$selectContrast_LS, 
                                       collapse = "+") %>%
                                    paste0("(",.,")/",length(input$selectContrast_LS)), 
                                sep = "-")
            vals$multipleCorrection_LS <- F
        } else {
            targetStage <- rbind(input$selectTarget_LS)
            contrastStage <- rbind(input$selectContrast_LS)
            comparison <- paste(input$selectTarget_LS, 
                                input$selectContrast_LS, 
                                sep = "-")
            vals$multipleCorrection_LS <- F
        }
        
        vals$targetStage_LS <- targetStage 
        vals$contrastStage_LS <- contrastStage
        vals$limmacontrast_LS <- comparison
        vals$comparison_LS <- gsub("/[0-9]*","", comparison)
        
        
    })
    
    ### LS: Generate Responsive Selection for Life Stage to Display ----
    output$contrastDisplaySelection_LS <- renderUI({
        comparison <- parse_contrasts_LS()
        
        panel(
            heading = tagList(h4(shiny::icon("fas fa-filter"),
                                 "Pick Contrast to Display")),
            status = "default",
            pickerInput("displayedComparison_LS",
                        NULL, 
                        comparison,
                        options = list(style = 'btn btn-primary'))
        )
    })
    
    ### LS: Set Contrast Matrix and Fit the Linear Model ----
    set_linear_model_LS <- eventReactive(input$goLS,{
        limma_ranking(vals$limmacontrast_LS, 
                      vals$targetStage_LS, 
                      vals$contrastStage_LS, 
                      vals$multipleCorrection_LS, 
                      NA, 
                      vals, fit, 
                      v.DEGList.filtered.norm, 
                      adj.P.thresh, 
                      diffGenes.df)
        
        vals$list.highlight.tbl_LS <- vals$list.highlight.tbl
    })
    
    ### LS: Extract the Differentially Expressed Genes ----
    pull_DEGs_LS <- reactive({
        req(vals$comparison_LS)
        req(vals$list.highlight.tbl_LS)
        
        if (isTruthy(input$displayedComparison_LS)){
            vals$displayedComparison_LS <- match(input$displayedComparison_LS,
                                                 vals$comparison_LS, nomatch = 1)
        } else {vals$displayedComparison_LS <- 1}
        #### Volcano Plots
        vplot <- ggplot(vals$list.highlight.tbl_LS[[vals$displayedComparison_LS]]) +
            aes(y=-log10(BH.adj.P.Val), x=logFC) +
            geom_point(size=2) +
            geom_hline(yintercept = -log10(adj.P.thresh), 
                       linetype="longdash", 
                       colour="grey", 
                       size=1) + 
            geom_vline(xintercept = lfc.thresh, 
                       linetype="longdash", 
                       colour="#BE684D", 
                       size=1) +
            geom_vline(xintercept = -lfc.thresh, 
                       linetype="longdash", 
                       colour="#2C467A", 
                       size=1) +
            labs(title = paste0('Pairwise Comparison: ',
                                gsub('-',
                                     ' vs ',
                                     vals$comparison_LS[vals$displayedComparison_LS])),
                 subtitle = paste0("grey line: p = ",
                                   adj.P.thresh, "; colored lines: log-fold change =", lfc.thresh),
                 color = "GeneIDs") +
            theme_Publication() +
            theme(aspect.ratio=1/3)
        vplot
    })
    
    ### LS: Volcano Plot, Generate UI ----
    output$volcano_LS <- renderPlot({
        req(input$goLS)
        set_linear_model_LS()
        pull_DEGs_LS()
    })
    
    ## LS: Save Volcano Plot
    output$downloadVolcano_LS <- downloadHandler(
        filename = function(){
            paste(vals$comparison_LS[vals$displayedComparison_LS], '_',Sys.Date(),'.pdf', sep='')
        },
        content = function(file){
            pull_DEGs_LS()
            ggsave(file,width = 11, height = 8, units = "in", device = "pdf", useDingbats=FALSE)
        }
    )
    
    # output$hover_info_LS <- renderUI({
    #     req(vals$displayedComparison)
    #     pointer.df <- vals$list.highlight.df[[vals$displayedComparison]] %>%
    #         dplyr::mutate(log10.adj.P.Val = -log10(BH.adj.P.Val))
    #     
    #     hover <- input$plot_hover_LS
    #     point <- nearPoints(pointer.df, hover, 
    #                         xvar = "logFC",
    #                         yvar = "log10.adj.P.Val",
    #                         threshold = 5, maxpoints = 1, addDist = TRUE)
    #     if (nrow(point) == 0) return(NULL)
    #     
    #     
    #     # calculate point position INSIDE the image as percent of total dimensions
    #     # from left (horizontal) and from top (vertical)
    #     left_pct <- (hover$x - hover$domain$left) / (hover$domain$right - hover$domain$left)
    #     top_pct <- (hover$domain$top - hover$y) / (hover$domain$top - hover$domain$bottom)
    #     
    #     # calculate distance from left and bottom side of the picture in pixels
    #     left_px <- hover$range$left + left_pct * (hover$range$right - hover$range$left)
    #     top_px <- hover$range$top + top_pct * (hover$range$bottom - hover$range$top)
    #     
    #     # create style property fot tooltip
    #     # background color is set so tooltip is a bit transparent
    #     # z-index is set so we are sure are tooltip will be on top
    #     style <- paste0("position:absolute; z-index:100; background-color: rgba(245, 245, 245, 0.85); ",
    #                     "left:", left_px - 200 , "px; top:", top_px - 200, "px;")
    #     
    #     # actual tooltip created as wellPanel
    #     wellPanel(
    #         style = style,
    #         p(HTML(paste0("<b> GeneID: </b>", 
    #                       point$geneID, 
    #                       "<br/>", 
    #                       "<b> Log FC: </b>",
    #                       round(point$logFC,digits = 2),
    #                       "<br/>",
    #                       "<b> p-value: </b>",
    #                       format(point$BH.adj.P.Val, digits = 3, scientific = TRUE))))
    #     )
    # })
    
    #### LS: Data Table of Differentially Expressed Genes ----
    assemble_DEGs_LS <- reactive({
        req(vals$displayedComparison_LS)
        tS<- vals$targetStage_LS[vals$displayedComparison_LS,
                                 ][vals$targetStage_LS[vals$displayedComparison_LS,
                                                       ]!=""]
        cS<- vals$contrastStage_LS[vals$displayedComparison_LS,
                                   ][vals$contrastStage_LS[vals$displayedComparison_LS,
                                                           ]!=""]
        
        n_num_cols <- length(tS)*3 + length(cS)*3 + 5
        LS.datatable <- vals$list.highlight.tbl_LS[[vals$displayedComparison_LS]] %>%
            DT::datatable(rownames = FALSE,
                          caption = htmltools::tags$caption(
                              style = 'caption-side: top; text-align: left; color: black',
                              htmltools::tags$b('Differentially Expressed Genes in', 
                                                htmltools::tags$em('S. stercoralis'), 
                                                gsub('-',' vs ',vals$comparison_LS[vals$displayedComparison_LS])),
                              htmltools::tags$br(),
                              "Threshold: p < ",
                              adj.P.thresh, "; log-fold change > ",
                              lfc.thresh,
                              htmltools::tags$br(),
                              'Values = log2 counts per million'),
                          options = list(autoWidth = TRUE,
                                         scrollX = TRUE,
                                         scrollY = '300px',
                                         scrollCollapse = TRUE,
                                         order = list(n_num_cols-1, 
                                                      'desc'),
                                         searchHighlight = TRUE, 
                                         pageLength = 25, 
                                         lengthMenu = c("5",
                                                        "10",
                                                        "25",
                                                        "50",
                                                        "100"),
                                         columnDefs = list(
                                             list(
                                                 targets = ((n_num_cols + 
                                                                 1)),
                                                 render = JS(
                                                     "function(data, row) {",
                                                     "data.toExponential(1);",
                                                     "}")
                                             ),
                                             list(
                                                 targets = ((n_num_cols + 
                                                                 4):(n_num_cols + 
                                                                         5)),
                                                 render = JS(
                                                     "function(data, type, row, meta) {",
                                                     "return type === 'display' && data.length > 20 ?",
                                                     "'<span title=\"' + data + '\">' + data.substr(0, 20) + '...</span>' : data;",
                                                     "}")
                                             ),
                                             list(targets = "_all",
                                                  class="dt-right")
                                         )
                                         
                          )) 
        LS.datatable <- LS.datatable %>%
            DT::formatRound(columns=c(3:n_num_cols), 
                            digits=3)
        
        LS.datatable <- LS.datatable %>%
            DT::formatRound(columns=c(n_num_cols+2, 
                                      n_num_cols+8), 
                            digits=2)
        
        LS.datatable <- LS.datatable %>%
            DT::formatSignif(columns=c(n_num_cols+1), 
                             digits=3)
        
        LS.datatable
    })
    
    
    output$tbl_LS <- renderDT ({
        req(input$goLS)
        DEG.datatable_LS<-assemble_DEGs_LS()
        DEG.datatable_LS
    })
    
    # LS: Save Excel Tables with DEG Tables ----
    output$downloadbuttonLS <- renderUI({
        req(input$goLS)
        req(vals$comparison_LS)
        
        output$generate_excel_report_LS <- generate_excel_report(vals$comparison_LS, 
                                                                 vals$list.highlight.tbl_LS)
        downloadButton("generate_excel_report_LS",
                       "Download DEG Tables",
                       class = "btn-primary")
    })
    
    
    session$onSessionEnded(stopApp)
}

# Run the application 
shinyApp(ui = ui, server = server)
