# Stercoralis RNAseq Browser Shiny App

## ---- Libraries ----
suppressPackageStartupMessages({
    library(shiny)
    library(htmltools)
    library(shinydashboard)
    library(tidyverse)
    library(ggplot2)
    library(RColorBrewer)
    library(ggdendro)
    library(magrittr)
    library(factoextra)
    library(gridExtra)
    library(cowplot)
    library(dendextend)
    library(DT)
})

## ---- Background ----
# Load pre-calculated data generated by S_stercoralis_RNAseq_ReAnalysis.R
load(file = "./SsLifeStageComparisons_subset")
print("Data Loaded Successfully")

## ---- UI ----

ui <- dashboardPage(
    
    # Application title
    title =("Strongyloides stercoralis RNAseq Browser"),
    
    # Header ----
    header = dashboardHeader(disable = T),
    
    # Sidebar ----
    sidebar = dashboardSidebar(disable = T),
    
    # Body ----
    body = dashboardBody(
        
        fluidRow(
            ## User Input Options ----
            
                   shinydashboard::box(
                       title = tagList(shiny::icon("fas fa-sliders-h"), "Inputs"),
                       width = 3,
                       status = "warning",
                       
                       # Select Target Life Stage
                       selectInput("selectTarget",
                                   h4("Select Target"),
                                   choices = list("FLF",
                                                  "PF",
                                                  "iL3",
                                                  "iL3a",
                                                  "ppL1",
                                                  "ppL3",
                                                  "pfL1"),
                                   selected = "iL3"),
                       # Select Contrast Life Stage
                       selectInput("selectContrast",
                                   h4("Select Contrast"),
                                   choices = list("FLF",
                                                  "PF",
                                                  "iL3",
                                                  "iL3a",
                                                  "ppL1",
                                                  "ppL3",
                                                  "pfL1"),
                                   selected = "FLF")
                   
                   
            ),
            
            
            # Reactive Outputs ----
            
                   shinydashboard::box(
                       title = tagList(shiny::icon("fas fa-calculator"),
                                       "Pairwise Differential Gene Expression Table"), 
                       width = 9,
                       status = "success",
                       side = "right",
                       DTOutput('tbl')
                   )
        )
    )
    
)


# Define server logic required to draw a histogram
server <- function(input, output, session) {
    
    vals <- reactiveValues(DEGtbl = NULL)
    
    output$tbl <- renderDT ({
        targetStage <- input$selectTarget
        contrastStage <- input$selectContrast
        comparison <- paste(targetStage, contrastStage, sep = "-")
        vals$DEGtbl <- DEGs.only[[comparison]]
        vals$DEGtbl
    })
    
    session$onSessionEnded(stopApp)
}

# Run the application 
shinyApp(ui = ui, server = server)
