# Stercoralis RNAseq Browser Shiny App

## ---- Libraries ----
suppressPackageStartupMessages({
    library(shiny)
    library(shinyWidgets)
    library(htmltools)
    library(shinydashboard)
    library(DT)
    library(tidyverse)
    library(limma) # differential gene expression using linear modeling
    library(edgeR)
    library(gt) 
    library(plotly)
    library(shinycssloaders)
    library(magrittr)
    source("Server/calc_DEG_tbl.R")
})

## ---- Background ----
# Load pre-calculated data generated by S_stercoralis_RNAseq_ReAnalysis.R
#load(file = "./SsLifeStageComparisons_subset")
#load(file = "./SsLifeStageComparisons_TopHits")
# Import a variance-stabilized DEGList created by voom transformation command.
# Outputs: E = normalized CPMexpression values on the log2 scale
load(file = "./vDEGList")

# Parse vDEGList into a tibble containing Log2CPM information
Log2CPM<-v.DEGList.filtered.norm$E %>%
    as_tibble(rownames = "geneID")%>%
    setNames(nm = c("geneID", as.character(v.DEGList.filtered.norm$targets$group))) %>%
    pivot_longer(cols = -geneID,
                 names_to = "life_stage", 
                 values_to = "log2CPM") %>%
    group_by(geneID, life_stage)

diffGenes.df <- v.DEGList.filtered.norm$E %>%
    as_tibble(rownames = "geneID", .name_repair = "unique")

# Fit a linear model to the data ----
fit <- lmFit(v.DEGList.filtered.norm, v.DEGList.filtered.norm$design)


# Set Expression threshold values for plotting and saving DEGs ----
adj.P.thresh <- 0.05
lfc.thresh <- 0

## ---- UI ----

# Application title ----
#

ui <- fluidPage(
    titlePanel("Strongyloides stercoralis RNAseq Browser"),
    source('UI/SsRNAseqBrowser_dashboard-ui.R', local = T)$value
    
)


# Define server logic  ----
server <- function(input, output, session) {
    
    vals <- reactiveValues(DEGtbl = NULL, 
                           genelist = NULL,
                           gene_of_interest = NULL,
                           highlight.df = NULL,
                           comparison = NULL,
                           targetStage = NULL,
                           contrastStage = NULL,
                           multipleCorrection = NULL,
                           results = NULL,
                           ebFit = NULL)
    
    ## GW Expression Plot ----
    parse_ids <- eventReactive(input$goGene,{
        if (isTruthy(input$idtext)){
            genelist <- input$idtext %>%
                gsub(" ", "", ., fixed = TRUE) %>%
                str_split(pattern = ",") %>%
                unlist() %>%
                as_tibble_col(column_name = "geneID")
        } else if (isTruthy(input$loadfile)){
            file <- input$loadfile
            ext <- tools::file_ext(file$datapath)
            validate(need(ext == "csv", "Please upload a csv file"))
            genelist <- read.csv(file$datapath, 
                                 header = FALSE, 
                                 colClasses = "character", 
                                 strip.white = T) %>%
                as_tibble() %>%
                pivot_longer(cols = everything(), values_to = "geneID") %>%
                dplyr::select(geneID)
        } 
        
        vals$genelist <- genelist
        
        # Select gene to look at expression levels
        vals$gene_of_interest <- vals$genelist[[1,1]]
    })
    
    ## Output: GW Expression Plot ----
    output$CPM <- renderPlot({
        parse_ids()
        
        gene_vals <- Log2CPM %>%
            dplyr::filter(geneID == vals$gene_of_interest)
        
        p<- ggplot(gene_vals) + 
            aes(x = life_stage, y = log2CPM, fill = life_stage) +
            geom_boxplot(show.legend = F, alpha = 0.7) +
            labs(y="log2 CPM expression", x = "Life Stage",
                 title=paste("Log2 Counts per Million (CPM):", vals$gene_of_interest),
                 subtitle="filtered, normalized, variance-stabilized") +
            theme_bw() +
            scale_fill_brewer(palette = "Dark2")
        p
    })
    
    ## Life-stage comparison ----
    parse_contrasts <- eventReactive(input$goGeneLS,{
        if (isTruthy(input$multiContrasts)) {
            comparison <- input$multiContrasts %>%
                gsub(" ", "", ., fixed = TRUE) %>%
                str_split(pattern = ",") %>%
                unlist()
            targetStage<- comparison %>%
                str_split(pattern="-", simplify = T) %>%
                .[,1] %>%
                gsub("(", "", ., fixed = TRUE) %>%
                gsub(")", "", ., fixed = TRUE) %>%
                str_split(pattern = "\\+", simplify = T)
            contrastStage<-comparison %>%
                str_split(pattern="-", simplify = T) %>%
                .[,2] %>%
                gsub("(", "", ., fixed = TRUE) %>%
                gsub(")", "", ., fixed = TRUE) %>%
                str_split(pattern = "\\+", simplify = T)
            vals$multipleCorrection <- T
        } else if (length(input$selectTarget_GW) > 1 | length(input$selectContrast_GW) > 1) {
            targetStage <- rbind(input$selectTarget_GW)
            contrastStage <- rbind(input$selectContrast_GW)
            comparison <- paste(paste0(input$selectTarget_GW, 
                                       collapse = "+") %>%
                                    paste0("(",.,")"), 
                                paste0(input$selectContrast_GW, 
                                       collapse = "+") %>%
                                    paste0("(",.,")"), 
                                sep = "-")
            vals$multipleCorrection <- F
        } else {
            targetStage <- rbind(input$selectTarget_GW)
            contrastStage <- rbind(input$selectContrast_GW)
            comparison <- paste(input$selectTarget_GW, input$selectContrast_GW, sep = "-")
            vals$multipleCorrection <- F
        }
        
        vals$targetStage <- targetStage 
        vals$contrastStage <- contrastStage
        vals$comparison <- comparison
        
    })
    
    output$contrastDisplaySelection <- renderUI({
        comparison <- parse_contrasts()
        selectInput("displayedComparison","Choose Comparison", comparison)
    })
    
    set_linear_model <- eventReactive(input$goGeneLS,{
        contrast.matrix <- makeContrasts(contrasts = vals$comparison,
                                         levels = v.DEGList.filtered.norm$design)
        
        ## Linear model fit -----
        fits <- contrasts.fit(fit, contrast.matrix)
        vals$ebFit <- limma::eBayes(fits)
        # Adjust for multiple comparisons using method = global. 
        
        if (vals$multipleCorrection == T) {
            vals$results <- decideTests(vals$ebFit, 
                                        method="global", 
                                        adjust.method="BH", 
                                        p.value = adj.P.thresh) %>%
                .[,] != 0
        }
        
    })
    
    pull_DEGs <- reactive({
        req(vals$genelist)
        
        if (isTruthy(input$displayedComparison)){
            vals$displayedComparison <- match(input$displayedComparison, vals$comparison)
        } else {vals$displayedComparison <- 1}
        
        myTopHits.df <- calc_DEG_tbl(vals$ebFit, vals$displayedComparison)
        
        # Filter dataset looking for the genes on the list ----
        highlight.df <- myTopHits.df %>%
            dplyr::filter(geneID %in% vals$genelist[[1]]) %>%
            dplyr::select(geneID, logFC, BH.adj.P.Val:percent_homology)
        vals$highlight.df <- highlight.df
        
        ## Volcano Plots ----
        vplot <- ggplot(myTopHits.df) +
            aes(y=-log10(BH.adj.P.Val), x=logFC) +
            geom_point(size=2) +
            geom_point(data = highlight.df, 
                       aes(y=-log10(BH.adj.P.Val), 
                           x=logFC, 
                           color = geneID, 
                           size = 3)) +
            geom_hline(yintercept = -log10(adj.P.thresh), linetype="longdash", colour="grey", size=1) + 
            geom_vline(xintercept = lfc.thresh, linetype="longdash", colour="#BE684D", size=1) +
            geom_vline(xintercept = -lfc.thresh, linetype="longdash", colour="#2C467A", size=1) +
            guides(size = FALSE) +
            labs(title = paste0(vals$comparison[vals$displayedComparison]),
                 color = "GeneIDs") +
            theme_bw()
        vplot
    })
    
    ## Outputs: Gene-wise Mode ----
    output$volcano <- renderPlot({
        set_linear_model()
        v <- pull_DEGs()
        v
    })
    
    ## GW Data Table ----
    assemble_DEGs_GW <- reactive({
        
        
        tS<- vals$targetStage[vals$displayedComparison,][vals$targetStage[vals$displayedComparison,]!=""]
        cS<- vals$contrastStage[vals$displayedComparison,][vals$contrastStage[vals$displayedComparison,]!=""]
        
        
        # Get log2CPM values for genes of interest
        highlight.tbl <- diffGenes.df %>%
            dplyr::filter(geneID %in% vals$genelist[[1]]) %>%
            dplyr::select(geneID, starts_with(paste0(tS,"-")), 
                          starts_with(paste0(cS,"-"))) %>%
            left_join(vals$highlight.df, by = "geneID")
        
        highlight.tbl$BH.adj.P.Val <- formatC(highlight.tbl$BH.adj.P.Val, digits = 3, format = "E") 
        
        
        n_num_cols <- length(tS)*3 + length(cS)*3 + 2
        #browser()
        highlight.datatable <- highlight.tbl %>%
            DT::datatable(extensions = c('KeyTable', "FixedHeader"),
                          rownames = FALSE,
                          caption = htmltools::tags$caption(
                              style = 'caption-side: top; text-align: left;',
                              htmltools::tags$b('Differentially Expressed Genes in', htmltools::tags$em('S. stercoralis'), 
                                                vals$comparison[vals$displayedComparison]),
                              htmltools::tags$br(),
                              "Threshold: p < ",
                              adj.P.thresh, "; log-fold change > ",
                              lfc.thresh,
                              htmltools::tags$br(),
                              'Values = log2 counts per million'),
                          options = list(keys = TRUE,
                                         autoWidth = TRUE,
                                         scrollX = TRUE,
                                         order = list(n_num_cols-1, 'desc'),
                                         searchHighlight = TRUE, 
                                         pageLength = 10, 
                                         lengthMenu = c("10", "25", "50", "100"),
                                         columnDefs = list(
                                             list(
                                                 targets = ((n_num_cols + 4):(n_num_cols + 5)),
                                                 render = JS(
                                                     "function(data, type, row, meta) {",
                                                     "return type === 'display' && data.length > 20 ?",
                                                     "'<span title=\"' + data + '\">' + data.substr(0, 20) + '...</span>' : data;",
                                                     "}")
                                             ),
                                             list(targets = "_all",
                                                  class="dt-right")
                                         )
                                         
                          )) 
        #browser()
        highlight.datatable <- highlight.datatable %>%
            DT::formatRound(columns=c(2:n_num_cols,n_num_cols+2, n_num_cols+8), digits=2)
        
        highlight.datatable
    })
    
    
    output$highlight.df <- renderDT ({
        #req(input$goGeneLS)
        req(vals$highlight.df)
        highlight.datatable<-assemble_DEGs_GW()
        highlight.datatable
    })
    
    
    
    ## Life stage-wise display of pairwise comparisons ----
    assemble_DEGs<-eventReactive(input$goLifeStage,{
        req((input$selectTarget != input$selectContrast))
        
        if (length(input$selectTarget > 1) | length(input$selectContrast > 1)){
            targetStage <- paste0(input$selectTarget, collapse = "+") %>%
                paste0("(",.,")")
            contrastStage <- paste0(input$selectContrast, collapse = "+") %>%
                paste0("(",.,")")
            comparison <- paste(targetStage, contrastStage, sep = "-")
        } else {
            targetStage <- input$selectTarget
            contrastStage <- input$selectContrast
            comparison <- paste(targetStage, contrastStage, sep = "-")
        }
        
        source('Server/assemble_DEGs.R', local = TRUE)$value
    })
    
    output$tbl <- renderDT ({
        req(input$goLifeStage)
        DEG.datatable<-assemble_DEGs()
        DEG.datatable
        
    })
    
    session$onSessionEnded(stopApp)
}

# Run the application 
shinyApp(ui = ui, server = server)
