# Strongyloides_RNAseq_Browser 
Web-based Shiny App for browsing and on-demand analysis of *Strongyloides spp* RNAseq datasets 

## Table of Contents  
1. [General Information](#general-information)
2. [App Setup & Deployment](#app-setup-&-deployment)
3. [App Features](#app-features)
4. [Examples of App Use](#examples-of-shiny-app-ui)
5. [Sources](#sources)
6. [License](#license)
7. [Authors](#authors)

## General Information
This repository contains non-responsive code for the pre-processing and analysis for *Strongyloides spp* RNAseq datasets. Preprocessed data as used as inputs into a Shiny Web App for online analysis and browsing of the data sets. 

The sections below describe the contents of the primary subfolders within this repository.

### Data  
This folder contains data files used in offline data pre-processing and analysis steps, including study design files and raw transcripts per million datasets for each species, as well as an Ensembl Compara database of Parasite Gene Sets and files containing previously published differential gene expression results that are used for benchmarking.

### Offline_Analysis  
This folder contains RMarkdown files for each species, in which we present example analyses. These analyses include hierarchical clustering and principal component analyses of species samples, limma-voom differential expression (example result echoing analyses completed by the Shiny App), benchmarking of differential expression analyses against previously published datasets, and an example gene set enrichment analysis. Rmd files are knitted into html files, and both PCA and Benchmarking plots are saved in a Plot subfolder in the Outputs folder.See the readme file located in the subfolder for additional details.

### Outputs  
This folder collects outputs generated by RMarkdown files located in the Offline_Analysis and Pre_processing folders.

### Pre_processing  
This folder contains RMarkdown files for each species; these files contain code for the alignment of raw reads, data filtering and normalization, voom variance-normalization of count data, and collection of gene annotation information. RMarkdown files generate outputs that act as essential inputs to the *Strongyloides* RNAseq Browser App. RMarkdown files are knitted into PDF files; see those PDF files for plots illustrating the effect of filtering and normallization on raw data inputs. See the readme file located in the subfolder for additional details.  

### Time_course_DGE  
This folder contains an RMarkdown file and cache that tests for differential expression on longitudinal data using the ImpulseDE2 package. The ImpulseDE2 method contrasts with DE algorithms such as limma that treat time points independently. Instead, ImpulseDE2 seeks to identify genes that display specific trajectories of differential gene expression over time. An Impulse model is designed to capture 4 different expression trajectories: monotonous decrease, monotonous increase, transient decrease (valley), and transient increase (peak). The analysis included in this folder specifically tests an implementation of ImpulseDE2 analysis on a subset of the *S. stercoralis* dataset: life stages FLF, PF, iL3, iL3a.  

### *Strongyloides*_RNAseq_Browser_App  
This folder contains the infrastructure for the Shiny application itself. This app is deployed via Shinyapps.io but can also be run locally. See App Setup and App Features sections below for additional details.  

## App Setup & Deployment
To access a stable deployment of the *Strongyloides* RNAseq Browser Web App, please visit:   [asbryant.shinyapps.io/strongyloides_rnaseq_browser/](asbryant.shinyapps.io/strongyloides_rnaseq_browser/)  


To run the latest version locally from Github, use the following command in R/RStudio:  
`library(shiny)`  
`shiny::runGitHub(repo = 'Strongyloides_RNAseq_Browser', username = 'astrasb', subdir = 'Strongyloides_RNAseq_Browser_App')`  

To run a specific release locally use the following commands in R/RStudio:  
  * For PCs --  
    `library(shiny)`  
    `shiny::runUrl('https://github.com/astrasb/Strongyloides_RNAseq_Browser/archive/<RELEASE_VERSION>.zip', subdir = 'Strongyloides_RNAseq_Browser_App/') ` 

  * For Macs --  
    `library(shiny)`  
    `shiny::runUrl('https://github.com/astrasb/Strongyloides_RNAseq_Browser/archive/<RELEASE_VERSION>.tar.gz', subdir = 'Strongyloides_RNAseq_Browser_App/')`  

Please note: the download step for runURL/runGitHub takes a substantial amount of time. We recommend downloading this archive and running the application locally. Note that full app functionality only requires the *Strongyloides*_RNAseq_Browser_App subdirectory.

## App Features  
The *Strongyloides* RNAseq Shiny App enables users to browse publically available *Strongyloides* bulk RNAseq datasets and perform on-demand analyses including differential expression and gene set enrichment. Data from the following species are currently included: *S. stercoralis*, *S. ratti*, *S. papillosus*, and *S. venezuelensis*. The app permits browsing RNAseq data in two modes:

  1. Browse by Life Stage Mode
  2. Browse by Gene Mode  
  
Features of the app include:  

* Search for gene(s) of interest using stable geneIDs or keywords
* Extract gene expression values for genes of interest
  - Display gene expresion across life stages as a heatmap (all genes of interest) or a boxplot (individual genes)
  - Download Log2 counts per million expression for genes of interest as .xslx
* On demand limma::voom-based pairwise differential gene expression analysis
  - Display results as interactive volcano plots and datatables
  - Download results as .pdf (plots) or .xlsx (datatables)
* Gene set enrichment analysis using the clusterProfiler R package and an Enss=embl Compara protein family set established by Hunt *et al* 2016. (Browse By Life Stage Mode)
  - Display results as bubble plot and interactive datatable
  - Download results as .pdf (plots) or .xlsx (datatables)
* Download raw/pre-processed data using user-friendly dropdown menu
  - Study desgin files (.csv)
  - Log2 counts per million expression for all genes and all samples (.csv)
  - Variance-stabilized DGEList object (R object; primary data input for the app)
  - Raw expression data for genes discarded during low-count filterering; see Pre-processing files for more information (.csv)

## Examples of Shiny App UI
![An example of the User Interface for the RNAseq Browser Shiny App in Browser by Gene Mode](./Strongyloides_RNAseq_Browser_App/UI/Str_RNAseq_Browser_Gene_Mode_Example.png)

![An example of the User Interface for the RNAseq Browser Shiny App in Browser by Life Stage Mode](./Strongyloides_RNAseq_Browser_App/UI/Str_RNAseq_Browser_LifeStage_Mode_Example.png)

## Sources
* [Shiny](https://shiny.rstudio.com/) - UI framework
* [kallisto](https://pachterlab.github.io/kallisto/) - Lightweight RNAseq pseudoalignment method
* [limma](https://bioconductor.org/packages/release/bioc/html/limma.html) - Diferential gene expression
* [clusterProfiler](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html) - Gene set enrichment analysis
* *Strongyloides* RNAseq datasets:
  - [Stolzfus *et al* 2012](https://journals.plos.org/plosntds/article?id=10.1371/journal.pntd.0001854)
  - [Hunt *et al* 2016](https://www.nature.com/articles/ng.3495)
  - [Hunt *et al* 2018](https://www.nature.com/articles/s41598-018-23514-z)
* [WormbaseParasite](https://parasite.wormbase.org/index.html) - Gene annotations and reference transcriptomes
* [DIYTranscriptomics](http://diytranscriptomics.com/) - Virtual asynchronous course where the authors learned best practices for RNAseq data analysis; provided primary pipeline for data pre-processing and analysis

## License  
This project is licensed under the MIT License. 

## Authors  
* [Astra Bryant, PhD](https://github.com/astrasb)
* [Stephanie DeMarco, PhD](https://github.com/sfdemarco)
