# Strongyloides_RNAseq_Browser 
Offline and Shiny App for *Strongyloides spp* RNAseq analysis 

## Table of Contents  
1. [General Information](#general-information)
2. [App Setup](#app-setup)
3. [App Features](#app-features)
4. [Examples of App Use](#example-of-shiny-app-ui)
5. [Sources](#sources)

## General Information
This repository contains non-responsive code for the pre-processing and analysis for *Strongyloides spp*  RNAseq datasets.  
Preprocessed data as used as inputs into a Shiny Web App for online analysis and browsing of the data sets. 

Use the sections below to view overviews of the primary subfolders within this repository.

### Data  
This folder contains data files used in offline data pre-processing and analysis steps, including study design files and raw transcripts per million datasets for each species, as well as an Ensembl Compara database of Parasite Gene Sets and files containing previously published differential gene expression results that are used for benchmarking.

### Offline_Analysis  
This folder contains RMarkdown files for each species, in which we present example analyses. These analyses include hierarchical clustering and principal component analyses of species samples, limma-voom differential expression (example result echoing analyses completed by the Shiny App), benchmarking of differential expression analyses against previously published datasets, and an example gene set enrichment analysis. Rmd files are knitted into html files, and both PCA and Benchmarking plots are saved in a Plot subfolder in the Outputs folder. See the Offline Example Analyses section below for additional details.

### Outputs  
This folder collects outputs generated by RMarkdown files located in the Offline_Analysis and Pre_processing folders.

### Pre_processing  
This folder contains RMarkdown files for each species; these files contain code for the alignment of raw reads, data filtering and normalization, voom variance-normalization of count data, and collection of gene annotation information. RMarkdown files generate outputs that act as essential inputs to the *Strongyloides* RNAseq Browser App. RMarkdown files are knitted into PDF files; see those PDF files for plots illustrating the effect of filtering and normallization on raw data inputs. See the Data Pre-processing section below for details on the how we process raw read data.

### *Strongyloides_RNAseq_Browser_App  
This folder contains the infrastructure for a Shiny application that permits browsing and analysis of RNAseq datasets. This app is deployed via Shinyapps.io but can also be run locally. See App Setup and App Features sections below for additional details.  

## App Setup
## App Features

## Example of Shiny App UI
![An example of the User Interface for the RNAseq Browser Shiny App](https://github.com/astrasb/Strongyloides_RNAseq_Browser/blob/master/Strongyloides_RNAseq_Browser_App/UI/St-RNAseq_Browser%20UI%20Example.png)

## Sources
