# Offline Analysis Overview  
This folder contains RMarkdown files for each species, in which we present example analyses. These analyses include hierarchical clustering and principal component analyses of species samples, limma-voom differential expression (example result echoing analyses completed by the Shiny App), benchmarking of differential expression analyses against previously published datasets, and an example gene set enrichment analysis. Rmd files are knitted into html files, and both PCA and Benchmarking plots are saved in a Plot subfolder in the Outputs folder.  

See sections below for descriptions of the analyses included in the .rmd files.  

## Table of Contents  
1. [Overview of RMD Files](#overview-of-rmd-files)
1. [Clustering Analysis](#clustering-analysis)
2. [Differential Gene Expression](#example-of-differential-gene-expression-analysis)
3. [Benchmarking](#benchmarking)
4. [Functional Enrichment Analysis](#functional-enrichment-analysis)

## Overview of RMD Files  
For all species, the offline Data Analysis RMarkdown file contains cluster analyses, examples of differential gene expression analyses conducted by the *Strongyloides* RNAseq Browser, and benchmarking comparisons against previously published datasets. Data files used as inputs for these analyses include the `SxRNAseq_data_preprocessed` and `Sx_vDGEList` R objects generated by the Pre-processing scripts. The code and results generated by the RMarkdown files are knitted into html files.  

## Clustering Analysis
The RMarkdown files first implement hierarchical clustering and PCA analyses on the data. Heirarchical clustering performed on filtered and normalized abundance data using the "complete" method. Identified principal components undergo additional analysis in order to determine which genes are influencing the identified PCs; an annotated list of genes that are the 10% of contributors (in any direction) to PC1 and PC2 is printed. Various graphs are plotted, including a dendrogram of the heirachical clustering, and several plots of visualize the PCA. PCA plots are saved in the Output folder (in the Plots subfolder).  

Next, we visualze gene expression across life stages by generating a heatmap for all the genes using variance-stabilized Log2CPM values. For heatmaps, columns (life stages) are ordered using Spearman clustering of expression in all genes (not just the user-defined subset). Rows are ordered using Pearson clustering. 

## Example of Differential Gene Expression Analysis
The `limma` package ( [Ritchie *et al* 2015](https://pubmed.ncbi.nlm.nih.gov/25605792/), [Phipson *et al* 2016](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5373812/)) is used to conduct pairwise differential gene expression analyses between life stages. For each species, an example pairwise comparison between two life stages is displayed as a volcano plot and interactive DataTable:

  * *S. stercoralis*: iL3 vs FLF  
  * *S. ratti*: iL3 vs FLF  
  * *S. papillosus*: iL3 vs FLF  
  * *S. venezuelensis*: PF vs FLF  

## Benchmarking
Next, we benchmark the results of our differential gene expression analysis against a published analsysis:

  * *S. stercoralis*: We use Supplementary Table 5 from [Hunt *et al* 2016](https://www.nature.com/articles/ng.3495), which includes the results of edgeR differential gene expression analysis between multiple life stages.  
  * *S. ratti*: We use Supplementary Table 5 from [Hunt *et al* 2016](https://www.nature.com/articles/ng.3495), which includes the results of edgeR differential gene expression analysis between multiple life stages.  
  * *S. papillosus*: Tables with data suitable for benchmarking are not currently available.  
  * *S. venezuelensis*: We use Supplementary Table 1 from [Hunt *et al* 2018](https://www.nature.com/articles/s41598-018-23514-z), which includes the results of edgeR differential gene expression analysis between free living females and parasitic females. Note that we manually adjust the geneID names in Table S1 in order to match current Wormbase ParaSite nomenclature, such that 'SVE_' is used instead of 'SVEN_'.  

## Functional Enrichment Analysis
Finally, we perform gene set enrichment analysis using the GSEA function from the `clusterProfiler` R package [(Yu et al 2012)](http://dx.doi.org/10.1089/omi.2011.0118). The goal of GSEA is to determine whether the members of a predefined set of genes are randomly distributed throughout a ranked gene list. Thus, GSEA depends on the availability of gene sets. Here, we use a Ensembl Compara protein family set list defined in [(Hunt *et al* 2016)](https://www.nature.com/articles/ng.3495). Note that the original set list uses specific transcript information, which we discard. The R script used to assemble the gene set list for use in this app is publically available as part of this repository (Pre_processing subfolder > `generateGeneSet.R`).  

For the example pairwise comparisons, the ranked gene list is constructed by rank ordering by LogFC. The GSEA analysis returns a plot of enriched gene families as well as a table containing normalized gene enrichment scores. These scores represent the degree to which the elements of the gene set are over-represented at the edges of the ranked gene list. Scores are normalized based on the number of genes within the gene set.

